### PR TITLE
Synthetic PoRep

### PR DESCRIPTION
**Pull request by cryptonemo**

### CODE REVIEW

File: storage-proofs-porep/src/stacked/vanilla/challenges.rs [hunk 1]
```diff
@@ -1,88 +1,536 @@
...
+
+use filecoin_hashers::Domain;
 use sha2::{Digest, Sha256};
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
+#[inline]
+fn bigint_to_challenge(bigint: BigUint, sector_nodes: usize) -> usize {
+    // Ensure that we don't challenge the first node.
+    let non_zero_node = (bigint % (sector_nodes - 1)) + 1usize;
+    // Assumes `sector_nodes` is less than 2^32.
```
Comment (Kubuxu) [inline storage-proofs-porep/src/stacked/vanilla/challenges.rs:15]: Maybe add debug assert?

File: storage-proofs-porep/src/stacked/vanilla/challenges.rs [hunk 2]
```diff
@@ -430,67 +361,42 @@ pub mod synthetic {
...
+            // The number of porep challenges generated per prf call.
+            let porep_challenges_per_block = 16;
+            // Indexes of chacha20 blocks used to generate partition `k`'s challenges.
+            let first_block = first_porep_challenge / porep_challenges_per_block;
+            let last_block = last_porep_challenge / porep_challenges_per_block;
+
+            // If the first porep challenge of the partition is not aligned to the first porep
+            // challenge generated by a chacha20 block, skip the leading challenges generated by the
+            // the first chacha20 block, i.e. skip the previous partition's last challenges.
+            let skip_first_challenges = first_porep_challenge % porep_challenges_per_block;
```
Comment (Kubuxu) [inline storage-proofs-porep/src/stacked/vanilla/challenges.rs]: We might be able to delete this as well if the chacha seeks allow you to seek in the middle of a block.

Comment (Kubuxu) [reply storage-proofs-porep/src/stacked/vanilla/challenges.rs]: If seeking to arbitrary byte position works for ChaCha here in Rust then you could seek to position: `first_porep_challenge*4` then `apply_keystream` to a buffer of size `4*num_partition_challenges`.

File: storage-proofs-porep/src/stacked/vanilla/params.rs [hunk 3]
```diff
@@ -304,6 +308,77 @@ impl<Tree: MerkleTreeTrait, G: Hasher> Proof<Tree, G> {
...
+    ) -> Self {
+        let challenge = col_proof.column.index as u64;
+        let num_layers = col_proof.column.rows.len();
+        let num_drg_parents = drg_col_proofs.len();
+        let num_exp_parents = exp_col_proofs.len();
+
+        let labeling_proofs: Vec<LabelingProof<Tree::Hasher>> = (0..num_layers)
+            .map(|layer_index| {
+                let has_exp_parents = (layer_index != 0) as usize;
+                let layer_parents = num_drg_parents + has_exp_parents * num_exp_parents;
```
Comment (vmx) [inline storage-proofs-porep/src/stacked/vanilla/params.rs:338]: Casting a boolean into a usize is pretty C-like. I'd prefer having it in a more type-safe way. What about:

```rust
let layer_parents = if layer_index == 0 {
    num_drg_parents
} else {
    num_drg_parents + num_exp_parents
}

**Review by vmx — COMMENTED** [review]
I've only reviewed some of the synthetic proofs serialization as I was interested in it for other reasons.

File: storage-proofs-core/src/merkle/proof.rs [hunk 4]
```diff
@@ -483,6 +494,65 @@ impl<
...
+        MerkleProof {
+            data: ProofData::Top(TopProof {
+                leaf,
+                root,
+                base_proof: base_path,
+                sub_proof: sub_path,
+                top_proof: top_path,
+            }),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-core/src/merkle/proof.rs:555]: Sure, this can be changed.

File: storage-proofs-porep/src/stacked/vanilla/params.rs (hunk 3 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-porep/src/stacked/vanilla/params.rs:338]: Yup, sounds good.

**Comment by jennijuju** [comment]
any reason to wait on landing the PR?

**Comment by cryptonemo** [comment]
> any reason to wait on landing the PR, as the test has ✅ ?

We'll have it merged sooner than later, but as it is now, all of synthetic-porep is in feature branches across all of the repos.  We didn't plan to cut a release until the SN work was complete (it's complete, but we're waiting on a release; @vmx has a meeting on that today and may have a solid update there).  After that, we're going to start thinking about our next release scheduling

**Comment by jennijuju** [comment]
> > any reason to wait on landing the PR, as the test has ✅ ?
> 
> We'll have it merged sooner than later, but as it is now, all of synthetic-porep is in feature branches across all of the repos. We didn't plan to cut a release until the SN work was complete (it's complete, but we're waiting on a release; @vmx has a meeting on that today and may have a solid update there). After that, we're going to start thinking about our next release scheduling

Thanks! why don’t we decouple the releases for synthetic PoRep and SN changes? From lotus perspective, they will be introduced in two separate releases, one in a mandatory consensus breaking release, the other in an optional feature release. Would be great if we can minimize the changes in proof libraries for the mandatory release (or external dependencies in general!).  (@rjan90 could you please follow up here after teams sync with SN and align with proof team on code landing/freeze expectations?)

**Pull request merged**